llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -540,6 +549,24 @@ struct BinaryOpc_match {
   }
 };
 
+/// Matches shuffle.
+template <typename T0, typename T1, typename T2> struct SDShuffle_match {
+  T0 Op1;
+  T1 Op2;
+  T2 Mask;
```

Comment:
Since `m_Mask` is different from other pattern classes -- it has `match(ArrayRef<int>)` rather than `match(Context, SDValue)`, I think you at least need to check if `T2` has the desired signature using type traits.
But more fundamentally, should be put another layer here? Could we just store the mask's `ArrayRef` directly here?

Reply:
Just implemented. I decided to go with the latter, storing the ArrayRef directly in `SDShuffle_match` as it seemed the simplest. I also replaced, in `SDShuffle_match`, the more generic `Mask.match(I->getMask())` with `std::equal(Mask.begin(), Mask.end(), I->getMask().begin())`, which was the same as the test coveraged I added in matchVecShuffle. Please let me know what you think of these changes.